File: src/tools/finance/conformal.ts
import {
  createForecastLabAssetScopedRuntimeDefaults,
  type ForecastLabRuntimeAssetScope,
} from './forecast-lab-runtime-defaults.js';

/**
 * Online Conformal PID Wrapper.
 *
 * Implements the PID-controller variant of online conformal prediction from
 * Angelopoulos, Candès & Tibshirani (2023), "Conformal PID Control for Time
 * Series Prediction", arXiv:2307.16895.
 *
 * The wrapper adapts a symmetric radius `q` so long-run miscoverage approaches
 * a target alpha. The integral decay γ < 1 prevents windup (§3 of the paper).
 */

export interface ConformalPIDOptions {
  /** Target miscoverage rate. 0.1 ⇒ 90% coverage. Default 0.1. */
  alpha?: number;
  /** Initial radius for the symmetric interval. Default 1.0. */
  initialRadius?: number;
  /** Base learning rate applied to the PID update. Default 0.05. */
  learningRate?: number;
  /** Proportional gain. Default 1.0. */
  kp?: number;
  /** Integral gain. Default 0.1. */
  ki?: number;
  /** Derivative gain. Default 0.1. */
  kd?: number;
  /** Decay applied to the integral term each step. Default 1.0 (no decay). */
  integralDecay?: number;
}

export interface ConformalInterval {
  low: number;
  high: number;
}

export type AdaptiveConformalMode = 'normal' | 'break';

export interface AdaptiveConformalRecordDiagnostics {
  structuralBreak?: boolean;
  realizedVol?: number;
}

export interface AdaptiveConformalPIDOptions extends ConformalPIDOptions {
  /**
   * Explicit on/off switch for structural-break-aware updates. Default false.
   */
  enabled?: boolean;
  /** Multiplier used for break-mode learning-rate and interval inflation. */
  breakLearningRateMultiplier?: number;
  /** Number of subsequent steps to keep break mode active after a trigger. */
  cooloffWindow?: number;
}

export interface AdaptiveConformalMetadata {
  applied: true;
  radius: number;
  coverageEstimate: number | null;
  mode: AdaptiveConformalMode;
}

export interface ScoreAggregatedConformalOptions {
  /** Target miscoverage rate. Reuses the conformal α. Default 0.1. */
  alpha?: number;
  /** Minimum calibration sample count before score aggregation activates. Default 20. */
  minSamples?: number;
  /** Maximum number of aggregated scores retained for calibration. Default 120. */
  calibrationWindow?: number;
}

export interface ScoreAggregatedConformalInterval extends ConformalInterval {
  applied: boolean;
  radius: number;
  multiplier: number | null;
}

export const FORECAST_LAB_CONFORMAL_PARAMETER_DEFAULTS: {
  readonly pidLearningRate: number;
  readonly integralDecay: number;
  readonly adaptiveBreakEnabled: boolean;
  readonly adaptiveBreakLearningRateMultiplier: number;
  readonly adaptiveBreakCooloffWindow: number;
  readonly scoreAggregationMinSamples: number;
  readonly scoreAggregationCalibrationWindow: number;
} = {
  pidLearningRate: 0.05,
  integralDecay: 1.0,
  adaptiveBreakEnabled: false,
  adaptiveBreakLearningRateMultiplier: 1.5,
  adaptiveBreakCooloffWindow: 0,
  scoreAggregationMinSamples: 16,
  scoreAggregationCalibrationWindow: 96,
};

const forecastLabConformalRuntimeDefaults = createForecastLabAssetScopedRuntimeDefaults(
  FORECAST_LAB_CONFORMAL_PARAMETER_DEFAULTS,
);

export function resolveForecastLabConformalParameterDefaults(
  assetScope?: ForecastLabRuntimeAssetScope,
): typeof FORECAST_LAB_CONFORMAL_PARAMETER_DEFAULTS {
  return forecastLabConformalRuntimeDefaults.resolve(assetScope);
}

export function getForecastLabConformalRuntimeDefaults(
  assetScope: ForecastLabRuntimeAssetScope,
): Partial<typeof FORECAST_LAB_CONFORMAL_PARAMETER_DEFAULTS> | undefined {
  return forecastLabConformalRuntimeDefaults.get(assetScope);
}

export function setForecastLabConformalRuntimeDefaults(
  assetScope: ForecastLabRuntimeAssetScope,
  overrides?: Partial<typeof FORECAST_LAB_CONFORMAL_PARAMETER_DEFAULTS>,
): void {
  forecastLabConformalRuntimeDefaults.set(assetScope, overrides);
}

export class ConformalPID {
  readonly alpha: number;
  readonly targetCoverage: number;
  protected readonly lr: number;
  protected readonly kp: number;
  protected readonly ki: number;
  protected readonly kd: number;
  protected readonly gamma: number;

  protected q: number;
  protected integral: number;
  protected prevBias: number;
  protected samples: number;
  protected hits: number;

  constructor(opts: ConformalPIDOptions = {}) {
    const defaults = resolveForecastLabConformalParameterDefaults();
    this.alpha = opts.alpha ?? 0.1;
    this.targetCoverage = 1 - this.alpha;
    this.q = Math.max(0, opts.initialRadius ?? 1.0);
    this.lr = opts.learningRate ?? defaults.pidLearningRate;
    this.kp = opts.kp ?? 1.0;
    this.ki = opts.ki ?? 0.1;
    this.kd = opts.kd ?? 0.1;
    this.gamma = opts.integralDecay ?? defaults.integralDecay;
    this.integral = 0;
    this.prevBias = 0;
    this.samples = 0;
    this.hits = 0;
  }

  /** Record a (forecastCenter, actual) pair and update the PID state. */
  record(forecastCenter: number, actual: number): void {
    this.step(forecastCenter, actual, this.lr);
  }

  protected step(
    forecastCenter: number,
    actual: number,
    learningRate: number,
  ): { residual: number; covered: boolean } | undefined {
    if (!Number.isFinite(forecastCenter) || !Number.isFinite(actual)) return;
    const residual = Math.abs(actual - forecastCenter);
    const covered = residual <= this.q ? 1 : 0;
    const err = 1 - covered; // 1 if missed, 0 if hit
    const bias = err - this.alpha;

    this.integral = this.gamma * this.integral + bias;
    const derivative = bias - this.prevBias;
    this.prevBias = bias;

    const update = learningRate * (this.kp * bias + this.ki * this.integral + this.kd * derivative);
    this.q = Math.max(0, this.q + update);

    this.samples += 1;
    this.hits += covered;
    return { residual, covered: covered === 1 };
  }

  /** Current symmetric radius. */
  currentRadius(): number {
    return this.q;
  }

  /** Reset controller state when the wrapped forecaster effectively restarts. */
  reset(opts: { radius?: number } = {}): void {
    if (opts.radius !== undefined) {
      this.q = Math.max(0, opts.radius);
    }
    this.integral = 0;
    this.prevBias = 0;
    this.samples = 0;
    this.hits = 0;
  }

  /** Wrap a forecast center in the current symmetric interval. */
  wrap(forecastCenter: number): ConformalInterval {
    return { low: forecastCenter - this.q, high: forecastCenter + this.q };
  }

  /** Number of samples observed so far. */
  sampleCount(): number {
    return this.samples;
  }

  /** Running empirical coverage. Undefined until the first record() call. */
  empiricalCoverage(): number | undefined {
    if (this.samples === 0) return undefined;
    return this.hits / this.samples;
  }
}

export class AdaptiveConformalPID extends ConformalPID {
  private readonly enabled: boolean;
  private readonly breakLearningRateMultiplier: number;
  private readonly cooloffWindow: number;
  private cooloffRemaining = 0;
  private residualEma?: number;
  private volatilityEma?: number;
  private mode: AdaptiveConformalMode = 'normal';
  private lastAppliedRadius: number;

  constructor(opts: AdaptiveConformalPIDOptions = {}) {
    super(opts);
    const defaults = resolveForecastLabConformalParameterDefaults();
    this.enabled = opts.enabled ?? defaults.adaptiveBreakEnabled;
    this.breakLearningRateMultiplier = Math.max(
      1,
      opts.breakLearningRateMultiplier ?? defaults.adaptiveBreakLearningRateMultiplier,
    );
    this.cooloffWindow = Math.max(
      0,
      Math.round(opts.cooloffWindow ?? defaults.adaptiveBreakCooloffWindow),
    );
    this.lastAppliedRadius = this.q;
  }

  override wrap(
    forecastCenter: number,
    diagnostics?: AdaptiveConformalRecordDiagnostics,
  ): ConformalInterval {
    const mode = this.resolveMode(diagnostics);
    const radius = this.appliedRadius(mode);
    this.lastAppliedRadius = radius;
    return { low: forecastCenter - radius, high: forecastCenter + radius };
  }

  override record(
    forecastCenter: number,
    actual: number,
    diagnostics?: AdaptiveConformalRecordDiagnostics,
  ): void {
    const mode = this.resolveMode(diagnostics, { consumeCooloff: true });
    const learningRate = mode === 'break'
      ? this.lr * this.breakLearningRateMultiplier
      : this.lr;
    const stepped = this.step(forecastCenter, actual, learningRate);
    if (!stepped) return;
    this.updateResidualEma(stepped.residual);
    this.updateVolatilityEma(diagnostics?.realizedVol);
  }

  currentMode(): AdaptiveConformalMode {
    return this.mode;
  }

  diagnostics(): AdaptiveConformalMetadata {
    return {
      applied: true,
      radius: this.lastAppliedRadius,
      coverageEstimate: this.empiricalCoverage() ?? null,
      mode: this.mode,
    };
  }

  override reset(opts: { radius?: number } = {}): void {
    super.reset(opts);
    this.cooloffRemaining = 0;
    this.residualEma = undefined;
    this.volatilityEma = undefined;
    this.mode = 'normal';
    this.lastAppliedRadius = this.q;
  }

  private resolveMode(
    diagnostics?: AdaptiveConformalRecordDiagnostics,
    options?: { consumeCooloff?: boolean },
  ): AdaptiveConformalMode {
    if (!this.enabled) {
      if (options?.consumeCooloff === true) this.cooloffRemaining = 0;
      this.mode = 'normal';
      return this.mode;
    }

    const triggered = diagnostics?.structuralBreak === true
      || this.isVolatilityShock(diagnostics?.realizedVol);
    const inCooloff = this.cooloffRemaining > 0;
    const mode = triggered || inCooloff ? 'break' : 'normal';

    if (options?.consumeCooloff === true) {
      if (triggered) {
        this.cooloffRemaining = this.cooloffWindow;
      } else if (inCooloff) {
        this.cooloffRemaining -= 1;
      }
    }

    this.mode = mode;
    return this.mode;
  }

  private appliedRadius(mode: AdaptiveConformalMode): number {
    if (mode !== 'break') return this.q;
    return this.q * Math.max(1, Math.sqrt(this.breakLearningRateMultiplier));
  }

  private isVolatilityShock(realizedVol?: number): boolean {
    if (!Number.isFinite(realizedVol) || realizedVol === undefined || realizedVol <= 0) return false;
    const baseline = this.volatilityEma ?? this.residualEma;
    if (baseline === undefined) return false;
    return realizedVol >= baseline * this.breakLearningRateMultiplier;
  }

  private updateResidualEma(residual: number): void {
    if (!Number.isFinite(residual) || residual < 0) return;
    this.residualEma = this.residualEma === undefined
      ? residual
      : (this.residualEma * 0.95) + (residual * 0.05);
  }

  private updateVolatilityEma(realizedVol?: number): void {
    if (!Number.isFinite(realizedVol) || realizedVol === undefined || realizedVol <= 0) return;
    this.volatilityEma = this.volatilityEma === undefined
      ? realizedVol
      : (this.volatilityEma * 0.9) + (realizedVol * 0.1);
  }
}

export class ScoreAggregatedConformal {
  private readonly alpha: number;
  private readonly minSamples: number;
  private readonly calibrationWindow: number;
  private readonly scores: number[] = [];

  constructor(opts: ScoreAggregatedConformalOptions = {}) {
    const defaults = resolveForecastLabConformalParameterDefaults();
    this.alpha = opts.alpha ?? 0.1;
    this.minSamples = Math.max(1, Math.round(opts.minSamples ?? defaults.scoreAggregationMinSamples));
    this.calibrationWindow = Math.max(
      this.minSamples,
      Math.round(opts.calibrationWindow ?? defaults.scoreAggregationCalibrationWindow),
    );
  }

  wrap(
    forecastCenter: number,
    sourceRadii: readonly number[],
  ): ScoreAggregatedConformalInterval {
    const radii = normalizeSourceRadii(sourceRadii);
    const baseRadius = radii.length > 0 ? Math.min(...radii) : 0;
    const multiplier = this.scoreMultiplier();
    const radius = baseRadius * (multiplier ?? 1);
    return {
      low: forecastCenter - radius,
      high: forecastCenter + radius,
      applied: multiplier !== undefined,
      radius,
      multiplier: multiplier ?? null,
    };
  }

  record(
    forecastCenter: number,
    actual: number,
    sourceRadii: readonly number[],
  ): void {
    if (!Number.isFinite(forecastCenter) || !Number.isFinite(actual)) return;
    const radii = normalizeSourceRadii(sourceRadii);
    if (radii.length === 0) return;

    const residual = Math.abs(actual - forecastCenter);
    const aggregatedScore = radii.reduce(
      (maxScore, radius) => Math.max(maxScore, residual / radius),
      0,
    );
    if (!Number.isFinite(aggregatedScore)) return;

    this.scores.push(aggregatedScore);
    if (this.scores.length > this.calibrationWindow) {
      this.scores.splice(0, this.scores.length - this.calibrationWindow);
    }
  }

  sampleCount(): number {
    return this.scores.length;
  }

  reset(): void {
    this.scores.length = 0;
  }

  private scoreMultiplier(): number | undefined {
    if (this.scores.length < this.minSamples) return undefined;
    return Math.max(1, upperQuantile(this.scores, 1 - this.alpha));
  }
}

function normalizeSourceRadii(sourceRadii: readonly number[]): number[] {
  return sourceRadii.filter((radius) => Number.isFinite(radius) && radius > 0);
}

function upperQuantile(values: readonly number[], quantile: number): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const clampedQuantile = Math.min(1, Math.max(0, quantile));
  const index = Math.min(
    sorted.length - 1,
    Math.max(0, Math.ceil(sorted.length * clampedQuantile) - 1),
  );
  return sorted[index];
}

File: research/models/conformal.py
"""Online Conformal PID wrapper (W3 Idea 2).

Mirrors ``src/tools/finance/conformal.ts``. Pure Python, no I/O.

Reference:
    Angelopoulos, Candès & Tibshirani (2023)
    "Conformal PID Control for Time Series Prediction"
    arXiv:2307.16895

The wrapper sits *outside* any forecasting model. Given a stream of
``(forecast_center, actual)`` pairs, it adapts a single radius ``q`` such that
the long-run miscoverage of the symmetric interval ``[center − q, center + q]``
approaches the target ``α``.  No assumptions on the underlying model — finite
sample coverage holds under arbitrary distribution shift.

The PID update is::

    bias_t  = err_t − α
    I_t     = γ · I_{t−1} + bias_t
    D_t     = bias_t − bias_{t−1}
    q_{t+1} = max(0, q_t + lr · (Kp·b + Ki·I + Kd·D))
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

from research.utils.forecast_lab_runtime_defaults import (
    create_forecast_lab_asset_scoped_runtime_defaults,
    ForecastLabRuntimeAssetScope,
)

FORECAST_LAB_CONFORMAL_PARAMETER_DEFAULTS: dict[str, float | int | bool] = {
    "pidLearningRate": 0.05,
    "integralDecay": 1.0,
    "adaptiveBreakEnabled": False,
    "adaptiveBreakLearningRateMultiplier": 1.5,
    "adaptiveBreakCooloffWindow": 0,
    "scoreAggregationMinSamples": 12,
    "scoreAggregationCalibrationWindow": 72,
}

PROMOTED_SOL_CONFORMAL_RUNTIME_DEFAULTS: dict[str, float | int | bool] = {
    "scoreAggregationMinSamples": 10,
    "scoreAggregationCalibrationWindow": 60,
}

_forecast_lab_conformal_runtime_defaults = create_forecast_lab_asset_scoped_runtime_defaults(
    FORECAST_LAB_CONFORMAL_PARAMETER_DEFAULTS
)
_forecast_lab_conformal_runtime_defaults.set("sol", PROMOTED_SOL_CONFORMAL_RUNTIME_DEFAULTS)


def resolve_forecast_lab_conformal_parameter_defaults(
    asset_scope: ForecastLabRuntimeAssetScope | None = None,
) -> dict[str, float | int | bool]:
    return _forecast_lab_conformal_runtime_defaults.resolve(asset_scope)


def get_forecast_lab_conformal_runtime_defaults(
    asset_scope: ForecastLabRuntimeAssetScope,
) -> dict[str, float | int | bool] | None:
    return _forecast_lab_conformal_runtime_defaults.get(asset_scope)


def set_forecast_lab_conformal_runtime_defaults(
    asset_scope: ForecastLabRuntimeAssetScope,
    overrides: dict[str, float | int | bool] | None = None,
) -> None:
    _forecast_lab_conformal_runtime_defaults.set(asset_scope, overrides)


@dataclass
class ConformalInterval:
    low: float
    high: float


class ConformalPID:
    """Online conformal prediction with a PID quantile controller."""

    def __init__(
        self,
        alpha: float = 0.1,
        initial_radius: float = 1.0,
        learning_rate: float | None = None,
        kp: float = 1.0,
        ki: float = 0.1,
        kd: float = 0.1,
        integral_decay: float | None = None,
    ) -> None:
        defaults = resolve_forecast_lab_conformal_parameter_defaults()
        self.alpha = float(alpha)
        self.target_coverage = 1.0 - self.alpha
        self.lr = float(
            learning_rate
            if learning_rate is not None
            else defaults["pidLearningRate"]
        )
        self.kp = float(kp)
        self.ki = float(ki)
        self.kd = float(kd)
        self.gamma = float(
            integral_decay
            if integral_decay is not None
            else defaults["integralDecay"]
        )
        self._initial_radius = max(0.0, float(initial_radius))
        self._q = self._initial_radius
        self._integral = 0.0
        self._prev_bias = 0.0
        self._samples = 0
        self._hits = 0

    # ------------------------------------------------------------------ #
    def _step(
        self, forecast_center: float, actual: float, learning_rate: float,
    ) -> dict | None:
        if not (math.isfinite(forecast_center) and math.isfinite(actual)):
            return None
        residual = abs(actual - forecast_center)
        covered = 1 if residual <= self._q else 0
        err = 1 - covered
        bias = err - self.alpha

        self._integral = self.gamma * self._integral + bias
        derivative = bias - self._prev_bias
        self._prev_bias = bias

        update = learning_rate * (
            self.kp * bias + self.ki * self._integral + self.kd * derivative
        )
        self._q = max(0.0, self._q + update)

        self._samples += 1
        self._hits += covered
        return {"residual": residual, "covered": covered == 1}

    def record(self, forecast_center: float, actual: float) -> None:
        self._step(forecast_center, actual, self.lr)

    # ------------------------------------------------------------------ #
    def current_radius(self) -> float:
        return self._q

    def reset(self, radius: Optional[float] = None) -> None:
        self._q = self._initial_radius if radius is None else max(0.0, float(radius))
        self._integral = 0.0
        self._prev_bias = 0.0
        self._samples = 0
        self._hits = 0

    def wrap(self, forecast_center: float) -> ConformalInterval:
        return ConformalInterval(
            low=forecast_center - self._q, high=forecast_center + self._q
        )

    def sample_count(self) -> int:
        return self._samples

    def empirical_coverage(self) -> Optional[float]:
        if self._samples == 0:
            return None
        return self._hits / self._samples


class AdaptiveConformalPID(ConformalPID):
    """Break-aware conformal PID wrapper matching the TypeScript behavior."""

    def __init__(
        self,
        alpha: float = 0.1,
        initial_radius: float = 1.0,
        learning_rate: float | None = None,
        kp: float = 1.0,
        ki: float = 0.1,
        kd: float = 0.1,
        integral_decay: float | None = None,
        enabled: bool | None = None,
        break_learning_rate_multiplier: float | None = None,
        cooloff_window: int | None = None,
    ) -> None:
        defaults = resolve_forecast_lab_conformal_parameter_defaults()
        super().__init__(
            alpha=alpha,
            initial_radius=initial_radius,
            learning_rate=learning_rate,
            kp=kp,
            ki=ki,
            kd=kd,
            integral_decay=integral_decay,
        )
        self._enabled = bool(
            defaults["adaptiveBreakEnabled"] if enabled is None else enabled
        )
        self._break_learning_rate_multiplier = max(
            1.0,
            float(
                defaults["adaptiveBreakLearningRateMultiplier"]
                if break_learning_rate_multiplier is None
                else break_learning_rate_multiplier
            ),
        )
        self._cooloff_window = max(
            0,
            int(
                round(
                    defaults["adaptiveBreakCooloffWindow"]
                    if cooloff_window is None
                    else cooloff_window
                )
            ),
        )
        self._cooloff_remaining = 0
        self._residual_ema: Optional[float] = None
        self._volatility_ema: Optional[float] = None
        self._mode = "normal"
        self._last_applied_radius = self._q

    def wrap(
        self,
        forecast_center: float,
        diagnostics: Optional[dict] = None,
    ) -> ConformalInterval:
        mode = self._resolve_mode(diagnostics)
        radius = self._applied_radius(mode)
        self._last_applied_radius = radius
        return ConformalInterval(
            low=forecast_center - radius,
            high=forecast_center + radius,
        )

    def record(
        self,
        forecast_center: float,
        actual: float,
        diagnostics: Optional[dict] = None,
    ) -> None:
        mode = self._resolve_mode(diagnostics, consume_cooloff=True)
        learning_rate = (
            self.lr * self._break_learning_rate_multiplier
            if mode == "break"
            else self.lr
        )
        stepped = self._step(forecast_center, actual, learning_rate)
        if stepped is None:
            return
        self._update_residual_ema(float(stepped["residual"]))
        self._update_volatility_ema(self._read_positive_float(diagnostics, "realized_vol"))

    def current_mode(self) -> str:
        return self._mode

    def diagnostics(self) -> dict:
        return {
            "applied": True,
            "radius": self._last_applied_radius,
            "coverage_estimate": self.empirical_coverage(),
            "mode": self._mode,
        }

    def reset(self, radius: Optional[float] = None) -> None:
        super().reset(radius=radius)
        self._cooloff_remaining = 0
        self._residual_ema = None
        self._volatility_ema = None
        self._mode = "normal"
        self._last_applied_radius = self._q

    def _resolve_mode(
        self,
        diagnostics: Optional[dict] = None,
        consume_cooloff: bool = False,
    ) -> str:
        if not self._enabled:
            if consume_cooloff:
                self._cooloff_remaining = 0
            self._mode = "normal"
            return self._mode

        triggered = (
            self._read_bool(diagnostics, "structural_break")
            or self._is_volatility_shock(
                self._read_positive_float(diagnostics, "realized_vol")
            )
        )
        in_cooloff = self._cooloff_remaining > 0
        mode = "break" if triggered or in_cooloff else "normal"

        if consume_cooloff:
            if triggered:
                self._cooloff_remaining = self._cooloff_window
            elif in_cooloff:
                self._cooloff_remaining -= 1

        self._mode = mode
        return self._mode

    def _applied_radius(self, mode: str) -> float:
        if mode != "break":
            return self._q
        return self._q * max(1.0, math.sqrt(self._break_learning_rate_multiplier))

    def _is_volatility_shock(self, realized_vol: Optional[float]) -> bool:
        if realized_vol is None or realized_vol <= 0:
            return False
        baseline = self._volatility_ema
        if baseline is None:
            baseline = self._residual_ema
        if baseline is None:
            return False
        return realized_vol >= baseline * self._break_learning_rate_multiplier

    def _update_residual_ema(self, residual: float) -> None:
        if not math.isfinite(residual) or residual < 0:
            return
        if self._residual_ema is None:
            self._residual_ema = residual
            return
        self._residual_ema = (self._residual_ema * 0.95) + (residual * 0.05)

    def _update_volatility_ema(self, realized_vol: Optional[float]) -> None:
        if realized_vol is None or not math.isfinite(realized_vol) or realized_vol <= 0:
            return
        if self._volatility_ema is None:
            self._volatility_ema = realized_vol
            return
        self._volatility_ema = (self._volatility_ema * 0.9) + (realized_vol * 0.1)

    @staticmethod
    def _read_bool(diagnostics: Optional[dict], key: str) -> bool:
        return bool(diagnostics and diagnostics.get(key) is True)

    @staticmethod
    def _read_positive_float(diagnostics: Optional[dict], key: str) -> Optional[float]:
        if not diagnostics:
            return None
        value = diagnostics.get(key)
        if isinstance(value, (int, float)) and math.isfinite(value) and value > 0:
            return float(value)
        return None


class ScoreAggregatedConformal:
    """Score-aggregation conformal wrapper matching TS ``ScoreAggregatedConformal``.

    Maintains a rolling window of nonconformity scores computed as the max
    of (residual / radius) across source radii, then calibrates a score-based
    quantile multiplier. Used in walk-forward backtesting when
    ``enableConformalScoreAggregation`` is true.
    """

    def __init__(
        self,
        alpha: float = 0.1,
        min_samples: int | None = None,
        calibration_window: int | None = None,
    ) -> None:
        defaults = resolve_forecast_lab_conformal_parameter_defaults()
        self._alpha = float(alpha)
        self._min_samples = max(
            1,
            int(
                round(
                    defaults["scoreAggregationMinSamples"]
                    if min_samples is None
                    else min_samples
                )
            ),
        )
        self._calibration_window = max(
            self._min_samples,
            int(
                round(
                    defaults["scoreAggregationCalibrationWindow"]
                    if calibration_window is None
                    else calibration_window
                )
            ),
        )
        self._scores: list[float] = []

    def wrap(
        self,
        forecast_center: float,
        source_radii: list[float],
    ) -> ConformalInterval:
        radii = _normalize_source_radii(source_radii)
        base_radius = min(radii) if radii else 0.0
        multiplier = self._score_multiplier()
        radius = base_radius * (multiplier if multiplier is not None else 1.0)
        return ConformalInterval(
            low=forecast_center - radius,
            high=forecast_center + radius,
        )

    def record(
        self,
        forecast_center: float,
        actual: float,
        source_radii: list[float],
    ) -> None:
        if not (math.isfinite(forecast_center) and math.isfinite(actual)):
            return
        radii = _normalize_source_radii(source_radii)
        if not radii:
            return

        residual = abs(actual - forecast_center)
        aggregated_score = max(
            (residual / radius for radius in radii),
            default=0.0,
        )
        if not math.isfinite(aggregated_score):
            return

        self._scores.append(aggregated_score)
        if len(self._scores) > self._calibration_window:
            self._scores = self._scores[-self._calibration_window:]

    def sample_count(self) -> int:
        return len(self._scores)

    def reset(self) -> None:
        self._scores.clear()

    def _score_multiplier(self) -> Optional[float]:
        if len(self._scores) < self._min_samples:
            return None
        return max(1.0, _upper_quantile(self._scores, 1.0 - self._alpha))


def _normalize_source_radii(source_radii: list[float]) -> list[float]:
    return [r for r in source_radii if math.isfinite(r) and r > 0]


def _upper_quantile(values: list[float], quantile: float) -> float:
    if not values:
        return 0.0
    sorted_vals = sorted(values)
    clamped = min(1.0, max(0.0, quantile))
    index = min(
        len(sorted_vals) - 1,
        max(0, int(math.ceil(len(sorted_vals) * clamped)) - 1),
    )
    return sorted_vals[index]
